packages/vincent-contracts/src/facets/VincentAppFacet.sol

```
@@ -0,0 +1,614 @@
+        if (toolIpfsCids.length == 0) {
+            revert NoToolsProvided(appId);
+        }
+
+        // Step 2: Validate input array lengths to ensure all tools have corresponding policies and parameters.
+        uint256 toolCount = toolIpfsCids.length;
+        if (
+            toolCount != toolPolicies.length || toolCount != toolPolicyParameterNames.length
+                || toolCount != toolPolicyParameterTypes.length
+        ) {
+            revert ToolsAndPoliciesLengthMismatch();
+        }
+
+        // Step 3: Fetch necessary storage references.
+        VincentAppStorage.AppStorage storage as_ = VincentAppStorage.appStorage();
+        VincentAppStorage.App storage app = as_.appIdToApp[appId];
+        VincentToolStorage.ToolStorage storage ts = VincentToolStorage.toolStorage();
+
+        // Step 4: Create a new app version.
+        app.versionedApps.push();
+        newAppVersion = app.versionedApps.length;
+
+        VincentAppStorage.VersionedApp storage versionedApp = app.versionedApps[getVersionedAppIndex(newAppVersion)];
+        versionedApp.enabled = true; // App versions are enabled by default
+
+        // Store this once outside the loop instead of repeatedly accessing it
+        EnumerableSet.Bytes32Set storage toolIpfsCidHashes = versionedApp.toolIpfsCidHashes;
+
+        // Step 5: Iterate through each tool to register it with the new app version.
+        for (uint256 i = 0; i < toolCount; i++) {
+            string memory toolIpfsCid = toolIpfsCids[i]; // Cache calldata value
+
+            // Validate tool IPFS CID is not empty
+            if (bytes(toolIpfsCid).length == 0) {
+                revert EmptyToolIpfsCidNotAllowed(appId, i);
+            }
+
+            bytes32 hashedToolCid = keccak256(abi.encodePacked(toolIpfsCid));
+
+            // Step 5.1: Register the tool IPFS CID globally if it hasn't been added already.
+            if (!toolIpfsCidHashes.contains(hashedToolCid)) {
```

What happens if you try to register the same tool twice: [Qm1, Qm1, Qm2]? We should throw an error

---

Also add a test for it we could have caught it earlier

---

Addressed by [this commit](https://github.com/LIT-Protocol/Vincent/pull/11/commits/7cd784e47aaab605fb9e54d164a51e8572cdaa22)